### ISSUE TITLE
Interactively using dart_eval

### ISSUE THREAD

**Issue by bingao** [ethanblake4/dart_eval#94]
Hi. I want to use dart_eval to develop an interactive interpreter, i.e. a user does not send codes to the interpreter at once. The interpreter should remember what (for example functions or variables) has been declared before when compiling and executing new codes from the user.

I wonder if it is possible? If yes, could you please provide some tips about what classes/methods I should use from dart_eval? Thank you.

**Comment by ethanblake4** [comment]
Hi! dart_eval isn't designed for this, so it could only be achieved by recompiling the entire source code the user has written so far each time a new statement is entered. That's more practical than it sounds since the dart_eval compiler is extremely fast (<100ms for a full recompile usually), but if you're targeting desktop platforms I'd look at https://github.com/fzyzcjy/dart_interactive instead.

**Comment by bingao** [comment]
Thank you @ethanblake4 I will look at `dart_interactive` but I am more interested in a library that can be used inside my app.

The problem I want to solve is some variables can take much time to compute from some functions. Users can reuse those variables later. So it would be better to collect those variables and let, for example, `Runtime` in `dart_eval` be aware of them instead of recomputing them. Do you think if it is possible for the current design of `dart_eval`? Thank you.

**Comment by ethanblake4** [comment]
No, not possible currently, though it wouldn't be terribly difficult to add. If you're interested in making a PR to add support for this I can give you some guidance.

**Comment by bingao** [comment]
Thanks. I am not sure if I can or currently have time to do it. But it is definitely good if you can give me some guidance :-)

**Comment by ethanblake4** [comment]
Sure, no problem.

The runtime values of top-level variables in dart_eval are stored in the [Runtime.globals](https://github.com/ethanblake4/dart_eval/blob/f95a589f0162487227c6900bb62cc12b7f1fd520/lib/src/eval/runtime/runtime.dart#L315) array.

The compiler keeps track of where each variable is in this array in [CompilerContext.topLevelGlobalIndices](https://github.com/ethanblake4/dart_eval/blob/f95a589f0162487227c6900bb62cc12b7f1fd520/lib/src/eval/compiler/compiler.dart#L43). For example, if you had a file `main.dart` with `final myVar = 1`, topLevelGlobalIndices might look like:
```dart
{
   'package:example/main.dart': {
     'myVar': 0
   }
}
```
And at runtime, the value of myVar (1) would be stored at index 0 of the Runtime.globals array.

In order to persist the state if you recompile, we'd want to copy over all values in the globals array from one Runtime to the next _if_ they are present in `topLevelGlobalIndices`. (We can't just copy the entire globals array because other things are stored there.)

In order to do this, you'd basically want to create an array out of all of the number values in topLevelGlobalIndices (so in the above example this would just be `[0]`). Then you'd need to stick this array inside of the generated `Program` [here](https://github.com/ethanblake4/dart_eval/blob/f95a589f0162487227c6900bb62cc12b7f1fd520/lib/src/eval/compiler/compiler.dart#L425), serialize it inside of [Program.write()](https://github.com/ethanblake4/dart_eval/blob/f95a589f0162487227c6900bb62cc12b7f1fd520/lib/src/eval/compiler/program.dart#L78), and deserialize it inside of [Runtime._load()](https://github.com/ethanblake4/dart_eval/blob/f95a589f0162487227c6900bb62cc12b7f1fd520/lib/src/eval/runtime/runtime.dart#L132). Finally you could add a new function on Runtime called `initializeGlobalStateFromPrevious` or similar, that takes in the previous Runtime as its argument and copies all the `globals` array indices that are present in the array.

Let me know if you have any questions!

**Comment by bingao** [comment]
Thank you @ethanblake4 I will have a try and see if I could make it work for my requirement.
